File: open_notebook/domain/scheduled_search.py
"""
ScheduledSearch domain model for recurring search automation.

Supports hourly, daily, weekly, monthly intervals.
Linked to notebooks and search engine providers.
"""

from datetime import datetime, timedelta, timezone
from typing import ClassVar, Optional

from open_notebook.domain.base import ObjectModel


INTERVAL_DELTAS = {
    "hourly": timedelta(hours=1),
    "daily": timedelta(days=1),
    "weekly": timedelta(weeks=1),
    "monthly": timedelta(days=30),
}


class ScheduledSearch(ObjectModel):
    table_name: ClassVar[str] = "scheduled_search"
    nullable_fields: ClassVar[set[str]] = {
        "last_run",
        "next_run",
        "last_error",
        "transformation_id",
    }

    name: str
    notebook_id: str
    query: str
    engine: str = "valyu"
    interval: str = "daily"  # hourly, daily, weekly, monthly
    is_active: bool = True
    last_run: Optional[datetime] = None
    next_run: Optional[datetime] = None
    run_count: int = 0
    last_error: Optional[str] = None
    transformation_id: Optional[str] = None
    save_as_source: bool = True

    def compute_next_run(self) -> datetime:
        """Compute the next run time based on interval."""
        delta = INTERVAL_DELTAS.get(self.interval, timedelta(days=1))
        now = datetime.now(timezone.utc)
        if self.last_run:
            return self.last_run + delta
        return now + delta

    async def mark_success(self) -> None:
        """Mark a successful run and schedule the next one."""
        self.last_run = datetime.now(timezone.utc)
        self.run_count += 1
        self.next_run = self.compute_next_run()
        self.last_error = None
        await self.save()

    async def mark_failure(self, error: str) -> None:
        """Record a failed run attempt."""
        self.last_run = datetime.now(timezone.utc)
        self.run_count += 1
        self.next_run = self.compute_next_run()
        self.last_error = error[:500]  # Truncate long errors
        await self.save()

    @classmethod
    async def get_due_searches(cls) -> list["ScheduledSearch"]:
        """Get all active scheduled searches that are due to run."""
        from open_notebook.database.repository import repo_query

        now = datetime.now(timezone.utc).isoformat()
        results = await repo_query(
            "SELECT * FROM scheduled_search WHERE is_active = true "
            "AND (next_run IS NONE OR next_run <= $now) "
            "ORDER BY next_run ASC",
            {"now": now},
        )
        return [cls(**r) for r in results]

    @classmethod
    async def get_by_notebook(cls, notebook_id: str) -> list["ScheduledSearch"]:
        """Get all scheduled searches for a specific notebook."""
        from open_notebook.database.repository import repo_query

        results = await repo_query(
            "SELECT * FROM scheduled_search WHERE notebook_id = $notebook_id "
            "ORDER BY created DESC",
            {"notebook_id": notebook_id},
        )
        return [cls(**r) for r in results]

File: frontend/src/lib/api/scheduled-search.ts
import apiClient from './client'

export interface ScheduledSearch {
  id: string
  name: string
  notebook_id: string
  query: string
  engine: string
  interval: string
  is_active: boolean
  last_run: string | null
  next_run: string | null
  run_count: number
  last_error: string | null
  transformation_id: string | null
  save_as_source: boolean
  created: string
  updated: string
}

export interface ScheduledSearchCreate {
  name: string
  notebook_id: string
  query: string
  engine?: string
  interval?: string
  transformation_id?: string
  save_as_source?: boolean
}

export interface ScheduledSearchUpdate {
  name?: string
  query?: string
  engine?: string
  interval?: string
  is_active?: boolean
  transformation_id?: string
  save_as_source?: boolean
}

export const scheduledSearchApi = {
  list: async (notebookId?: string): Promise<ScheduledSearch[]> => {
    const params = notebookId ? { notebook_id: notebookId } : {}
    const response = await apiClient.get<ScheduledSearch[]>('/scheduled-searches', { params })
    return response.data
  },

  get: async (id: string): Promise<ScheduledSearch> => {
    const response = await apiClient.get<ScheduledSearch>(`/scheduled-searches/${id}`)
    return response.data
  },

  create: async (data: ScheduledSearchCreate): Promise<ScheduledSearch> => {
    const response = await apiClient.post<ScheduledSearch>('/scheduled-searches', data)
    return response.data
  },

  update: async (id: string, data: ScheduledSearchUpdate): Promise<ScheduledSearch> => {
    const response = await apiClient.put<ScheduledSearch>(`/scheduled-searches/${id}`, data)
    return response.data
  },

  delete: async (id: string): Promise<{ message: string }> => {
    const response = await apiClient.delete<{ message: string }>(`/scheduled-searches/${id}`)
    return response.data
  },

  trigger: async (id: string): Promise<{ message: string; result: Record<string, unknown> }> => {
    const response = await apiClient.post<{ message: string; result: Record<string, unknown> }>(`/scheduled-searches/${id}/run`)
    return response.data
  },

  runDue: async (): Promise<{ message: string; results: Record<string, unknown>[] }> => {
    const response = await apiClient.post<{ message: string; results: Record<string, unknown>[] }>('/scheduled-searches/run-due')
    return response.data
  },
}
